### PR TITLE
.Net: IKernel - Pre/Post RunAsync Handlers - Part 1

### LINKED ISSUE #1714: Be able to Handle Function Calls: Phase 1
### IKernel Handlers - Notification

An IKernel.RunAsync Caller needs:

- Know and be notified before and after each function is executed with information it can get or change prior and after invocation.

  - Pre-Invocation

    - Get: Function details
    - Get: Prompt generated by the current Kernel `TemplateEngine` before calling the LLM
    - Get: Current settings used
    - Get: SKContext
      - Get: State of variables (args)

    - Set: Cancel execution
    - Set: Modify parameters (Context Variables)

  - Post-Invocation

    - Get: Function details
    - Get: Prompt generated by the current Kernel `TemplateEngine` before calling the LLM
    - Get: SKContext
      - Get: LLM Model Result (Tokens Usage, Stop Sequence, ...)
      - 
    - Set: Modify parameters (Context Variables)

### PR DESCRIPTION
**Pull request by RogerBarreto**
### Motivation and Context

Be able to add handlers to function executions using `Kernel.RunAsync` in a way the caller can be notified and get or change information before and after a function is executed.

Resolves #1714 

### Description

This implementation adds 2 new public interface events to a IKernel interface allowing setting Pre(Running)/Post(Ran) handlers to it:

- IKernel.FunctionInvoking
- IKernel.FunctionInvoked

With this initial change allow scenarios like:

- Pre-Execution (Invoking)

    - Get: SKContext
    - Set: Modify the SKContext variables
    - Set: Cancel further RunAsync function executions

- Post-Execution (Invoked)
    - Get: SKContext
    - Set: Modify the SKContext variables
    - Set: Cancel further RunAsync function executions

## Out of scope of this PR.

- Plan/Step Handlers execution support
- Individual ISKFunction.InvokeAsync execution support
- Get prompt in Pre/Post events

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#dev-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Events;
+using RepoUtils;
+
+#pragma warning disable RCS1214 // Unnecessary interpolated string.
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs:11]: Is this needed? I don't see any issues (also they should be easy to resolve, yeah?)

File: dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        await ChangingResultAsync();
+
+        await BeforeInvokeCancellationAsync();
+
+        await AfterInvokeCancellationAsync();
+
+        await SkippingFunctionsAsync();
+    }
+
+    private static async Task GetPromptAndUsageAsync()
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs:43]: nit: naming isn't accurate at the moment (no prompt)

File: dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+public static class Example57_FunctionEventHandlers
+{
+    private static string? openAIModelId;
+    private static string? openAIApiKey;
+
+    public static async Task RunAsync()
+    {
+        Console.WriteLine("\n======== Using Function Execution Handlers ========\n");
+
+        openAIModelId = TestConfiguration.OpenAI.ModelId;
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs:23]: nit: We should make it easier for our examples to leverage azure by default if available and then fallback to openai.

**Review by lemillermicrosoft — APPROVED** [review]

File: dotnet/src/SemanticKernel/Kernel.cs [hunk 4]
```diff
@@ -249,6 +278,38 @@ public void Dispose()
     private readonly ILogger _logger;
     private readonly IDelegatingHandlerFactory _httpHandlerFactory;
 
+    /// <summary>
+    /// Execute the FunctionInvoking event handlers and returns true if no cancellation was attempted.
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Kernel.cs:282]: `returns true...` - comments out of date?

**Review by shawncal — APPROVED** [review]

**Review by shawncal — APPROVED** [review]

**Pull request merged**